File: app/src/main/java/fuck/wifiadbtoggle/droidvendorssuck/NetworkMonitorService.java
package fuck.wifiadbtoggle.droidvendorssuck;

import android.app.Notification;
import android.app.Service;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.net.wifi.WifiManager;
import android.os.Build;
import android.os.IBinder;
import android.os.PowerManager;
import java.util.Set;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class NetworkMonitorService extends Service {

    private ConnectivityManager connectivityManager;
    private final ExecutorService executor = Executors.newSingleThreadExecutor();
    private PowerManager.WakeLock wakeLock = null;
    private boolean legacyReceiverRegistered = false;

    private final BroadcastReceiver legacyReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            scheduleEvaluate();
        }
    };

    @Override
    public void onCreate() {
        super.onCreate();
        connectivityManager = (ConnectivityManager) getSystemService(Context.CONNECTIVITY_SERVICE);
        startForeground(NotificationHelper.STATUS_NOTIF_ID, buildNotification());
        registerCallback();
        updateWakeLock();
        scheduleEvaluate();
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        if (!Settings.isAutoStartEnabled(this) || !Settings.isAnyMonitorRuleEnabled(this)) {
            stopSelf();
            return START_NOT_STICKY;
        }
        startForeground(NotificationHelper.STATUS_NOTIF_ID, buildNotification());
        updateWakeLock();
        scheduleEvaluate();
        return START_STICKY;
    }

    @Override
    public void onDestroy() {
        unregisterCallback();
        releaseWakeLock();
        executor.shutdown();
        super.onDestroy();
    }

    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    private void registerCallback() {
        try {
            IntentFilter filter = new IntentFilter(ConnectivityManager.CONNECTIVITY_ACTION);
            registerReceiver(legacyReceiver, filter);
            legacyReceiverRegistered = true;
        } catch (Exception ignored) {
            // Ignore registration failures
        }
    }

    private void unregisterCallback() {
        try {
            if (legacyReceiverRegistered) {
                unregisterReceiver(legacyReceiver);
                legacyReceiverRegistered = false;
            }
        } catch (Exception ignored) {
            // Ignore unregister failures
        }
    }

    private void scheduleEvaluate() {
        executor.execute(this::evaluateAndApply);
    }

    private void evaluateAndApply() {
        try {
            if (!Settings.isAutoStartEnabled(this) || !Settings.isAnyMonitorRuleEnabled(this)) {
                stopSelf();
                return;
            }
            if (!ShellRunner.canUseRoot()) return;
            ScheduleMode scheduleMode = ScheduleManager.getActiveMode(this);
            if (scheduleMode == ScheduleMode.FORCE_ON) {
                if (!AdbWifiController.isEnabled(this)) {
                    AdbWifiController.enable(this);
                }
                return;
            }
            if (scheduleMode == ScheduleMode.FORCE_OFF) {
                if (AdbWifiController.isEnabled(this)) {
                    AdbWifiController.disable(this);
                }
                return;
            }

            NetworkInfo info = connectivityManager.getActiveNetworkInfo();
            boolean isConnected = info != null && info.isConnected();
            boolean onWifi = isConnected && info.getType() == ConnectivityManager.TYPE_WIFI;
            boolean onEthernet = isConnected && info.getType() == ConnectivityManager.TYPE_ETHERNET;

            if (!onWifi && !onEthernet) {
                if (Settings.isDisableOnDisconnectEnabled(this)) {
                    if (AdbWifiController.isEnabled(this)) {
                        AdbWifiController.disable(this);
                    }
                }
                return;
            }

            if (onEthernet && Settings.isAutoEnableEthernetEnabled(this)) {
                if (!AdbWifiController.isEnabled(this)) {
                    AdbWifiController.enable(this);
                }
                return;
            }

            if (onWifi && Settings.isAutoEnableSsidEnabled(this)) {
                WifiInfoResult wifiInfo = getWifiInfo();
                String ssid = wifiInfo != null ? wifiInfo.ssid : null;
                String bssid = wifiInfo != null ? wifiInfo.bssid : null;
                if (ssid == null) return;
                Set<String> allowedSsids = Settings.getSsidSet(this);
                if (!allowedSsids.isEmpty() && !allowedSsids.contains(Settings.normalizeSsid(ssid))) return;
                if (Settings.isFilterBssidEnabled(this)) {
                    if (bssid == null) return;
                    Set<String> allowedBssids = Settings.getBssidSet(this);
                    if (!allowedBssids.isEmpty() && !allowedBssids.contains(Settings.normalizeBssid(bssid))) return;
                }
                if (!AdbWifiController.isEnabled(this)) {
                    AdbWifiController.enable(this);
                }
            }
        } finally {
            updateStatusNotification();
        }
    }

    private WifiInfoResult getWifiInfo() {
        try {
            WifiManager wifiManager = (WifiManager) getApplicationContext().getSystemService(Context.WIFI_SERVICE);
            if (wifiManager == null || wifiManager.getConnectionInfo() == null) return null;
            String ssid = wifiManager.getConnectionInfo().getSSID();
            if (ssid == null || "<unknown ssid>".equals(ssid)) return null;
            String bssid = wifiManager.getConnectionInfo().getBSSID();
            return new WifiInfoResult(ssid, bssid);
        } catch (Exception ignored) {
            return null;
        }
    }

    private void updateWakeLock() {
        if (Settings.isKeepAwakeEnabled(this)) {
            if (wakeLock == null || !wakeLock.isHeld()) {
                PowerManager pm = (PowerManager) getSystemService(Context.POWER_SERVICE);
                wakeLock = pm.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, getPackageName() + ":adb_monitor");
                wakeLock.acquire();
            }
        } else {
            releaseWakeLock();
        }
    }

    private void releaseWakeLock() {
        try {
            if (wakeLock != null && wakeLock.isHeld()) {
                wakeLock.release();
            }
        } catch (Exception ignored) {
            // Ignore
        } finally {
            wakeLock = null;
        }
    }

    private Notification buildNotification() {
        if (Settings.isPersistentNotificationEnabled(this)) {
            return NotificationHelper.buildStatusNotification(this);
        }
        return NotificationHelper.buildMonitorNotification(this);
    }

    public static void start(Context context) {
        if (!BuildConfig.FEATURE_MONITOR) return;
        if (!Settings.isAutoStartEnabled(context) || !Settings.isAnyMonitorRuleEnabled(context)) {
            return;
        }
        Intent intent = new Intent(context, NetworkMonitorService.class);
        if (Build.VERSION.SDK_INT >= 26) {
            context.startForegroundService(intent);
        } else {
            context.startService(intent);
        }
    }

    public static void stop(Context context) {
        Intent intent = new Intent(context, NetworkMonitorService.class);
        context.stopService(intent);
    }

    private void updateStatusNotification() {
        if (!Settings.isPersistentNotificationEnabled(this)) return;
        NotificationHelper.notifyStatus(this);
    }

    private static final class WifiInfoResult {
        final String ssid;
        final String bssid;

        WifiInfoResult(String ssid, String bssid) {
            this.ssid = ssid;
            this.bssid = bssid;
        }
    }
}

File: app/src/main/java/fuck/wifiadbtoggle/droidvendorssuck/NetworkMonitorService.kt
package fuck.wifiadbtoggle.droidvendorssuck

import android.app.Notification
import android.app.Service
import android.content.Context
import android.content.Intent
import android.net.ConnectivityManager
import android.net.wifi.WifiManager
import android.os.Build
import android.os.IBinder
import android.os.PowerManager
import androidx.core.content.ContextCompat
import java.util.concurrent.Executors

class NetworkMonitorService : Service() {

    private lateinit var connectivityManager: ConnectivityManager
    private val executor = Executors.newSingleThreadExecutor()
    private var wakeLock: PowerManager.WakeLock? = null
    private var legacyReceiverRegistered = false

    private val legacyReceiver = object : android.content.BroadcastReceiver() {
        override fun onReceive(context: Context?, intent: Intent?) {
            scheduleEvaluate()
        }
    }

    override fun onCreate() {
        super.onCreate()
        connectivityManager = getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        startForeground(NotificationHelper.STATUS_NOTIF_ID, buildNotification())
        registerCallback()
        updateWakeLock()
        scheduleEvaluate()
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        if (!Settings.isAutoStartEnabled(this) || !Settings.isAnyMonitorRuleEnabled(this)) {
            stopSelf()
            return START_NOT_STICKY
        }
        startForeground(NotificationHelper.STATUS_NOTIF_ID, buildNotification())
        updateWakeLock()
        scheduleEvaluate()
        return START_STICKY
    }

    override fun onDestroy() {
        unregisterCallback()
        releaseWakeLock()
        executor.shutdown()
        super.onDestroy()
    }

    override fun onBind(intent: Intent?): IBinder? = null

    private fun registerCallback() {
        try {
            val filter = android.content.IntentFilter(ConnectivityManager.CONNECTIVITY_ACTION)
            registerReceiver(legacyReceiver, filter)
            legacyReceiverRegistered = true
        } catch (_: Exception) {
            // Ignore registration failures
        }
    }

    private fun unregisterCallback() {
        try {
            if (legacyReceiverRegistered) {
                unregisterReceiver(legacyReceiver)
                legacyReceiverRegistered = false
            }
        } catch (_: Exception) {
            // Ignore unregister failures
        }
    }

    private fun scheduleEvaluate() {
        executor.execute { evaluateAndApply() }
    }

    private fun evaluateAndApply() {
        try {
            if (!Settings.isAutoStartEnabled(this) || !Settings.isAnyMonitorRuleEnabled(this)) {
                stopSelf()
                return
            }
            if (!ShellRunner.canUseRoot()) return
            val scheduleMode = ScheduleManager.getActiveMode(this)
            if (scheduleMode == ScheduleMode.FORCE_ON) {
                if (!AdbWifiController.isEnabled(this)) {
                    AdbWifiController.enable(this)
                }
                return
            }
            if (scheduleMode == ScheduleMode.FORCE_OFF) {
                if (AdbWifiController.isEnabled(this)) {
                    AdbWifiController.disable(this)
                }
                return
            }

            val info = connectivityManager.activeNetworkInfo
            val isConnected = info?.isConnected == true
            val onWifi = isConnected && info?.type == ConnectivityManager.TYPE_WIFI
            val onEthernet = isConnected && info?.type == ConnectivityManager.TYPE_ETHERNET

            if (!onWifi && !onEthernet) {
                if (Settings.isDisableOnDisconnectEnabled(this)) {
                    if (AdbWifiController.isEnabled(this)) {
                        AdbWifiController.disable(this)
                    }
                }
                return
            }

            if (onEthernet && Settings.isAutoEnableEthernetEnabled(this)) {
                if (!AdbWifiController.isEnabled(this)) {
                    AdbWifiController.enable(this)
                }
                return
            }

            if (onWifi && Settings.isAutoEnableSsidEnabled(this)) {
                val wifiInfo = getWifiInfo()
                val ssid = wifiInfo?.first
                val bssid = wifiInfo?.second
                if (ssid == null) return
                val allowedSsids = Settings.getSsidSet(this)
                if (allowedSsids.isNotEmpty() && !allowedSsids.contains(Settings.normalizeSsid(ssid))) return
                if (Settings.isFilterBssidEnabled(this)) {
                    if (bssid == null) return
                    val allowedBssids = Settings.getBssidSet(this)
                    if (allowedBssids.isNotEmpty() && !allowedBssids.contains(Settings.normalizeBssid(bssid))) return
                }
                if (!AdbWifiController.isEnabled(this)) {
                    AdbWifiController.enable(this)
                }
            }
        } finally {
            updateStatusNotification()
        }
    }

    private fun getWifiInfo(): Pair<String, String?>? {
        return try {
            val wifiManager = applicationContext.getSystemService(Context.WIFI_SERVICE) as WifiManager
            val info = wifiManager.connectionInfo ?: return null
            val ssid = info.ssid?.takeIf { it != "<unknown ssid>" } ?: return null
            ssid to info.bssid
        } catch (_: Exception) {
            null
        }
    }

    private fun updateWakeLock() {
        if (Settings.isKeepAwakeEnabled(this)) {
            if (wakeLock?.isHeld != true) {
                val pm = getSystemService(Context.POWER_SERVICE) as PowerManager
                wakeLock = pm.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, "$packageName:adb_monitor")
                wakeLock?.acquire()
            }
        } else {
            releaseWakeLock()
        }
    }

    private fun releaseWakeLock() {
        try {
            wakeLock?.takeIf { it.isHeld }?.release()
        } catch (_: Exception) {
            // Ignore
        } finally {
            wakeLock = null
        }
    }

    private fun buildNotification(): Notification {
        return if (Settings.isPersistentNotificationEnabled(this)) {
            NotificationHelper.buildStatusNotification(this)
        } else {
            NotificationHelper.buildMonitorNotification(this)
        }
    }

    companion object {
        fun start(context: Context) {
            if (!Settings.isAutoStartEnabled(context) || !Settings.isAnyMonitorRuleEnabled(context)) {
                return
            }
            val intent = Intent(context, NetworkMonitorService::class.java)
            if (Build.VERSION.SDK_INT >= 26) {
                ContextCompat.startForegroundService(context, intent)
            } else {
                context.startService(intent)
            }
        }

        fun stop(context: Context) {
            val intent = Intent(context, NetworkMonitorService::class.java)
            context.stopService(intent)
        }
    }

    private fun updateStatusNotification() {
        if (!Settings.isPersistentNotificationEnabled(this)) return
        NotificationHelper.notifyStatus(this)
    }
}
